lib/que/metrics.rb

```diff
@@ -0,0 +1,91 @@
+# frozen_string_literal: true
+
+require "rack"
+require "benchmark"
+require "prometheus/client"
+require "prometheus/middleware/exporter"
+
+module Que
+  class Metrics
+    def initialize(labels: {})
+      @registry = Prometheus::Client::Registry.new
+
+      @jobs_worked_total = @registry.counter(:que_jobs_worked_total, "Counter for all jobs processed", labels)
+      @jobs_worked_seconds_total = @registry.counter(:que_jobs_worked_seconds_total, "Sum of the time taken processing each job class", labels)
+      @jobs_latency_seconds_total = @registry.counter(:que_jobs_latency_seconds_total, "Sum of time spent by job and priority waiting in queue", labels)
+      @worker_running_seconds_total = @registry.counter(:que_worker_running_seconds_total, "Time since starting to work jobs", labels)
+      @worker_job_acquire_total = @registry.counter(:que_worker_job_acquire_total, "Counter of number of attempts to acquire jobs", labels)
+      @worker_job_acquire_seconds_total = @registry.counter(:que_worker_job_acquire_seconds_total, "Sum of time taken to acquire jobs", labels)
+    end
+
+    def expose(port: 8080)
+      Rack::Handler::WEBrick.run(
+        Prometheus::Middleware::Exporter.new(Proc.new {}, registry: @registry), Port: port,
+      )
+    end
+
+    def start_work
+      stop = false
+
+      Thread.new do
+        last_timestamp = monotonic_now
+
+        until stop
+          timestamp = monotonic_now
+          @worker_running_seconds_total.increment({}, timestamp - last_timestamp)
+          last_timestamp = timestamp
+
+          sleep(1) # delay the next update for 1s
+        end
+      end
+
+      -> { stop = true }
```

Comment:
I think you'd get slightly more accurate metrics if you set `timestamp` in this lambda.

```ruby
timestamp = nil
Thread.new do
  last_timestamp = monotonic_now
  until timestamp
    @worker_running_seconds_total.increment({}, timestamp - last_timestamp)
    sleep(1)
  end

-> { timestamp = monotonic_now }
```

Reply:
Yes, very true!

Reply:
But the `last_timestamp` is definitely not in scope in this lambda. I think I'm going to not do this, as I reckon a second doesn't matter in the grand scheme of things. Prometheus will be scraping this every 15s anyway, so we allow for <15s inaccuracies in these times, so 1s isn't going to bother us.